### PR TITLE
defer immediately

### PR DESCRIPTION
**Pull request by gelfand**

### CODE REVIEW

File: eth/stagedsync/stage_execute.go [hunk 1]
```diff
@@ -302,29 +299,31 @@ Loop:
...
-				if err != nil {
-					return err
+				if !useExternalTx {
+					if err := s.Update(tx, stageProgress); err != nil {
+						return err
+					}
+					if err := tx.Commit(); err != nil {
+						return err
+					}
+					tx, err := cfg.db.BeginRw(context.Background())
```
Comment (AskAlexSharov) [inline eth/stagedsync/stage_execute.go:313]: `tx` variable shadowed. At he end of execution stage you will call Commit on another variable.

File: eth/stagedsync/stage_execute.go [hunk 2]
```diff
@@ -302,29 +299,31 @@ Loop:
...
+					if err != nil {
+						return err
+					}
+					defer tx.Rollback()
```
Comment (AskAlexSharov) [inline eth/stagedsync/stage_execute.go:317]: immediate rollback of just created transaction - means not possible to write anything in it anymore

File: eth/stagedsync/stage_execute.go [hunk 3]
```diff
@@ -302,29 +299,31 @@ Loop:
...
 				}
-				// TODO: This creates stacked up deferrals
-				defer tx.Rollback()
+				batch = olddb.NewBatch(tx, quit)
+				defer batch.Rollback()
```
Comment (AskAlexSharov) [inline eth/stagedsync/stage_execute.go:320]: `defer batch.Rollback()` at the end of function - is equivalent to just `batch.Rollback()`. But it also means you create object and cancel it right on next line.

**Comment by gelfand** [comment]
Move #2670

**Pull request closed without merging**